pageserver/src/tenant/mgr.rs

```diff
@@ -1101,101 +1313,313 @@ pub(crate) async fn attach_tenant(
+                Err(TenantSlotUpsertError::InternalError(
+                    "Missing InProgress marker during tenant upsert".into(),
+                ))
+            }
+            Some(slot) => {
+                error!(tenant_id=%self.tenant_id, "Unexpected contents of TenantSlot during upsert, this is a bug.  Contents: {:?}", slot);
+                Err(TenantSlotUpsertError::InternalError(
+                    "Unexpected contents of TenantSlot".into(),
+                ))
+            }
+        }
     }
 }
 
-async fn tenant_map_upsert_slot<'a, F, R>(
-    tenant_id: TenantId,
-    upsert_fn: F,
-) -> Result<(), TenantMapInsertError>
-where
-    F: FnOnce(Option<TenantSlot>) -> R,
-    R: std::future::Future<Output = anyhow::Result<TenantSlot>>,
-{
-    let mut guard = TENANTS.write().await;
-    let m = match &mut *guard {
-        TenantsMap::Initializing => return Err(TenantMapInsertError::StillInitializing),
-        TenantsMap::ShuttingDown(_) => return Err(TenantMapInsertError::ShuttingDown),
+impl Drop for SlotGuard {
+    fn drop(&mut self) {
+        if !self.upserted {
+            let mut locked = TENANTS.write().unwrap();
+
+            let m = match &mut *locked {
+                TenantsMap::Initializing | TenantsMap::ShuttingDown(_) => {
+                    return;
+                }
+                TenantsMap::Open(m) => m,
+            };
+
+            let slot = m.remove(&self.tenant_id);
+            match slot {
+                Some(slot) => match slot {
+                    TenantSlot::InProgress(_) => {
+                        // Normal case: nothing should have replaced the TenantSlot value
+                        // that was set when we were constructed.
+                    }
+                    _ => {
+                        error!(tenant_id=%self.tenant_id, "Unexpected contents of TenantSlot during upsert, this is a bug.  Contents: {:?}", slot);
+                    }
+                },
+                None => {
+                    error!(
+                        tenant_id = %self.tenant_id,
+                        "Missing InProgress marker during SlotGuard drop, this is a bug."
+                    );
+                }
+            }
+        }
+    }
+}
+
+/// `allow_shutdown=true` is appropriate for read-only APIs that should stay working while
+/// the pageserver is shutting down.  Otherwise, set it to false to refuse attempts to
+/// operate on a slot while we are shutting down.
+///
+fn tenant_map_peek_slot<'a>(
+    tenants: &'a std::sync::RwLockReadGuard<'a, TenantsMap>,
+    tenant_id: &TenantId,
+    allow_shutdown: bool,
+) -> Result<Option<&'a TenantSlot>, TenantMapError> {
+    let m = match tenants.deref() {
+        TenantsMap::Initializing => return Err(TenantMapError::StillInitializing),
+        TenantsMap::ShuttingDown(m) => {
+            if allow_shutdown {
+                return Err(TenantMapError::ShuttingDown);
+            } else {
+                m
+            }
```

Comment:
This doesn't match my understanding of the semantics in the doc comment. My understanding from the doc comment is that one should pass `true` when one is ok to operate on a tenant while we're shutting down.

Reply:
Fixed in 810404f9